core/src/main/java/com/graphhopper/util/InstructionList.java

```diff
@@ -342,4 +342,86 @@ public Instruction find(double lat, double lon, double maxDistance) {
         return get(foundInstruction);
     }
 
+    /**
+     * Update the InstructionList using the the available context and order of instructions to improve the instructions.
+     * If the requests contains a heading, this method can transform the first continue to a u-turn if the heading
+     * points into the opposite direction of the route.
+     * At a waypoint it can transform the continue to a u-turn if the route involves turning.
+     * Consecutive continue instruction can be merged if they are close and use the same street name.
+     *
+     * @param favoredHeading The favored heading of the request, which can create an initial u-turn
+     */
+    public void updateInstructionsWithContext(double favoredHeading) {
+        Instruction instruction;
+        Instruction nextInstruction;
+
+        final int MERGE_CONTINUE_INSTRUCTIONS_DISTANCE = 40;
+
+        for (int i = 0; i < this.size() - 1; i++) {
+            instruction = this.get(i);
+            nextInstruction = this.get(i + 1);
+
+            if (i == 0 && !Double.isNaN(favoredHeading) && instruction.getExtraInfoJSON().containsKey("heading")) {
+                double heading = (double) instruction.getExtraInfoJSON().get("heading");
```

Comment:
I do not like using something internally that we have implemented for the JSON export. Instead we should use normal Java variables with compile time checks etc or implement this in a different way.

Probably we should create a java variable for the (last_)heading and also think about creating a JSON serializer for the Instruction class to replace InstructionList.createJson

Reply:
Actually this is not really part of this PR, this was added in #1073, where this parameter was added for this and for the JSON, [here](https://github.com/graphhopper/graphhopper/pull/1073/files#diff-e5132d18a6606ee07a176bff362f8ccfR129). IIRC, the reason why I implemented it this way instead of a regular property was that we only use the heading for the first continue instruction. We add the `last_heading` to all FinishInstructions. This allows us to compare the first and last heading.

We could introduce a new Instruction subclass for the initial Instruction and store the heading property there. We could store the last_heading in the FinishInstruction. We could also add the heading to every instruction, but I don't think that this would make much sense.

Should we move this to a separate issue?

Reply:
Ok, I see that this is already used. Let us get this fixed first in another PR as this is too unclean and afterwards proceed here?

Regarding the different headings: why couldn't we always use `heading` why do we need `last_heading`? The finish instruction would then get the heading of the very last edge.